### PR TITLE
Add log profiler support to Mono MSVC build.

### PR DESCRIPTION
**Pull request by lateralusX**
Several fixes to the log profiler in order to work on Windows.

* Add new dynamic library project to build log profiler on Windows MSVC.
* Fix dependecy between log profiler and Mono.
* Rewrite command pipe logic on Windows since it's not supported to select
  on both sockets and pipes.
* Several additional socket fixes needed for log profiler to work on Windows.

NOTE, this makes sure we have a log profiler that can be started/stopped and at least generate some basic output. Still needs more testing on Windows since there might be other parts of the profiler not working as expected (since it has not been run for quite some time).

### CODE REVIEW

File: mono/profiler/log.c [hunk 1]
```diff
@@ -2862,13 +2881,63 @@ cleanup_reusable_samples (void)
...
+				int max_fd = -1;
+
+				FD_ZERO (&wfds);
+				FD_SET (client_socket, &wfds);
+
+				/*
+				* Include timeout to prevent hanging on connect call.
+				*/
+				struct timeval tv = { .tv_sec = 1, .tv_usec = 0 };
+				select (client_socket + 1, NULL, &wfds, NULL, &tv);
```
Comment (jaykrell) [inline mono/profiler/log.c]: Designated initializers generally removed from tree for  C++-compat.

Comment (jaykrell) [reply mono/profiler/log.c]: WaitForMultipleObjects likely could use pipes & sockets together.

Comment (lateralusX) [reply mono/profiler/log.c]: To make that work reliable you would need to move over to WSA API's and then you will need to use event's an implement that logic together with network events to get something similar. Didn't want to go in to much depth doing that, just fixing this command side channel, only used during shutdown, only passing a command that is represented as a char. Current change will keep most of the logic as is.

**Comment by lateralusX** [comment]
@monojenkins rebuild failed

File: mono/profiler/log.c (hunk 1 above)
Comment (lateralusX) [reply mono/profiler/log.c]: The initializer was a copy from the other select in this source, so we should probably fix both. Maybe that could be done in separate PR.

Comment (jaykrell) [reply mono/profiler/log.c]: Fair enough. I have ported this directory to C++ at least once before..

Comment (jaykrell) [reply mono/profiler/log.c]: Btw I don't think you need to use WSA for wait, I think just regular WaitForMultipleObjects will do.
Regular sockets are in kernel and it just works.
Sockets can be implemented I guess via a usermode provider, but those are then wrapped up by a kernel driver to restore kernel wait functions to working. As I understand.

Comment (jaykrell) [reply mono/profiler/log.c]: But fair enough to both points.

**Comment by jaykrell** [comment]
Notice the implementation of `WSAWaitForMultipleEvents`:.
```
C:\s>\bin\amd64\cdb curl

0:000> u ws2_32!WSAWaitForMultipleEvents

WS2_32!WSAWaitForMultipleEvents:
00007ffd`00443e90 48ff2579200300  jmp     qword ptr [WS2_32!_imp_WaitForMultipleObjectsEx (00007ffd`00475f10)]
```

**Comment by lateralusX** [comment]
I was not primarly thinking about the wait method but setting up similar pattern using it, most documentation points to WSASelectEvent + WSAEnumNetworkEvents for this or switch to IO completion ports. Mixing that with anonymous pipes was not worth the effort and didn't want to change it that much (and I wanted to limit impact on other platforms as well).

**Comment by akoeplinger** [comment]
@lateralusX I guess we should add the new library to the Windows MSI ?

**Comment by lateralusX** [comment]
@akoeplinger  I believe we could, file is called, mono-profiler-log.dll.

**Pull request merged**